src/celllistimpl.cpp

```diff
@@ -0,0 +1,263 @@
+    const int forward_neighbors_count = neighbors.size() / 2;
+    forward_neighbors.clear();
+    forward_neighbors.reserve(forward_neighbors_count);
+    std::copy_if(neighbors.begin(), neighbors.end(), std::back_inserter(forward_neighbors), [](auto& offset) {
+        return (offset > CellCoord{0, 0, 0}).all();
+    });
+    assert(forward_neighbors.size() == forward_neighbors_count);
+}
+
+/**
+ * Tests
+ */
+TEST_CASE("Grid3DFixed") {
+    using TestGridBase = Grid3DFixed;
+    using CellCoord = TestGridBase::CellCoord;
+    TestGridBase grid({20., 30., 40.}, 1.5);
+    CellCoord end{(200 / 15), (300 / 15), (400 / 15)};
+    CellCoord first{0, 0, 0};
+    CellCoord last = end - CellCoord{1, 1, 1};
+    CellCoord coord1{5, 6, 7};
+
+    REQUIRE_EQ(grid.size(), end.prod());
+
+    SUBCASE("Cell coordinates are limited to [origin, last]") {
+        CHECK(grid.isCell(first));
+        CHECK_FALSE(grid.isCell(first - CellCoord{1, 0, 0}));
+        CHECK_FALSE(grid.isCell(first - CellCoord{0, 1, 0}));
+        CHECK_FALSE(grid.isCell(first - CellCoord{0, 0, 1}));
+
+        CHECK(grid.isCell(last));
+        CHECK_FALSE(grid.isCell(last + CellCoord{1, 0, 0}));
+        CHECK_FALSE(grid.isCell(last + CellCoord{0, 1, 0}));
+        CHECK_FALSE(grid.isCell(last + CellCoord{0, 0, 1}));
+
+        CHECK_FALSE(grid.isCell(end));
+
+        CHECK(grid.isCell(coord1));
+        CHECK(grid.isCell(2 * coord1));
+        CHECK_FALSE(grid.isCell(4 * coord1));
+
+        CHECK_EQ(grid.index(first), 0);
+        CHECK_EQ(grid.index(last), grid.size() - 1);
+
+        CHECK((grid.coordinates(grid.index(coord1)) == coord1).all());
+        CHECK_NE(grid.index(coord1.reverse()), grid.index(coord1));
+
+        SUBCASE("Offsets cannot wrap around") {
+            CHECK(grid.isNeighborCell(first, {0, 0, 0}));
+            CHECK(grid.isNeighborCell(first, {0, 0, 1}));
+            CHECK_FALSE(grid.isNeighborCell(first, {0, 0, -1}));
+            CHECK(grid.isNeighborCell(last, {0, 0, 0}));
+            CHECK(grid.isNeighborCell(last, {0, 0, -1}));
+            CHECK_FALSE(grid.isNeighborCell(last, {0, 0, 1}));
+        }
+
+        SUBCASE("Spatial coordinates has to be inside the box") {
+            CHECK(grid.isCellAt({0., 0., 0.}));
+            CHECK(grid.isCellAt({1., 1., 1.}));
+            CHECK(grid.isCellAt({19., 29., 39.}));
+            CHECK_FALSE(grid.isCellAt({20., 30., 40.}));
+
+            CHECK((grid.coordinatesAt({1., 1., 1.}) == first).all());
+            CHECK((grid.coordinatesAt({19., 29., 39.}) == last).all());
+
+            CHECK((grid.coordinatesAt({3., 2.5, 3.}) == CellCoord{1, 1, 1}).all());
+            CHECK((grid.coordinatesAt({3., 3., 3.}) == CellCoord{1, 2, 1}).all());
+        }
+    }
+}
+
+TEST_CASE("Grid3DPeriodic") {
+    using TestGridBase = Grid3DPeriodic;
+    using CellCoord = TestGridBase::CellCoord;
+    TestGridBase grid({20., 30., 40.}, 1.5);
+    CellCoord end{(200 / 15), (300 / 15), (400 / 15)};
```

Comment:
integer division intended?

Reply:
Yes. Cell coordinates are an ordered tuple of integers (`CellCoord` is `Vector3i`).